dnsdist: add support for labels on custom metrics

esensar opened the pull request:
### Short description

This adds support for prometheus labels on custom metrics. Changes the metrics maps to hold a map of labels to metric values for each metric. Metrics without labels have an empty string label combination value.

Adds an optional options table as the last parameter of `incMetric`, `decMetric`, `setMetric` and `getMetric`, which can be used to set labels by using the `label` key or to set the previously used parameter (either `step` or `value`).

Closes: #13359

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

rgacogne commented on pdns/dnsdistdist/dnsdist-metrics.cc:
```diff
@@ -176,18 +178,16 @@ std::optional<std::string> declareCustomMetric(const std::string& name, const st
-    auto itp = customCounters->insert({name, MutableCounter()});
+    auto itp = customCounters->emplace(name, std::map<std::string, MutableCounter>());
     if (itp.second) {
-      g_stats.entries.write_lock()->emplace_back(Stats::EntryPair{name, &(*customCounters)[name].d_value});
```
One problem with this change is that a custom metric that is never set/incremented/decremented doesn't show up in the metrics anymore. That breaks our regression tests, but more importantly I think it breaks some user expectations. Do you think it would make sense to add an optional parameter (perhaps we should make it an optional LuaAssociativeTable right away so we don't end up with too many parameters) to this method to specify whether the metrics is expected to have labels? Metrics without labels (the default since it is the existing behaviour) would be added to `g_stats` right away, and metrics with labels wouldn't.

esensar replied:
Alright, that makes sense. I just removed it since it caused issues for ones with labels, but I didn't expect to break the regular use case :smile: 
I assume we can make the old behavior default and then only skip this step if explicitly marked as a metric with labels.

I agree about the optional LuaAssociativeTable, since all of the other metrics related functions have these now.

rgacogne commented on pdns/dnsdistdist/dnsdist-metrics.cc:
```diff
@@ -198,57 +198,108 @@ std::optional<std::string> declareCustomMetric(const std::string& name, const st
   if (metric != customCounters->end()) {
-    metric->second.d_value -= step;
-    return metric->second.d_value.load();
+    auto combinationOfLabels = generateCombinationOfLabels(labels);
+    auto metricEntry = metric->second.find(combinationOfLabels);
+    if (metricEntry == metric->second.end()) {
+      metricEntry = metric->second.emplace(combinationOfLabels, MutableCounter()).first;
+      g_stats.entries.write_lock()->emplace_back(Stats::EntryTriple{std::string(name), combinationOfLabels, &metricEntry->second.d_value});
+    }
+    metricEntry->second.d_value -= step;
+    return metricEntry->second.d_value.load();
```
Just a nit but it looks like we could get rid of a bit of code duplication by turning this into a (templated to handle both counters and gauges?) function.

esensar replied:
Thank you! I will take a shot and let you know if I get stuck

rgacogne reviewed (changes requested):
I made a few comments, and I think we need to exclude metrics with labels from both Carbon export and `dumpStats()`, but I'm quite happy with the general concept, and the code itself is good as well.

esensar commented:
By the way, an unrelated question:
I have tried to fix formatting in files I have touched, using the provided script (`./build-scripts/format-code...`). Doing that has produced wrong results, which failed on CI. On further inspection I saw that the script expects `clang-format-11`, while I had `clang-format-17` installed. After installing 11, it worked as expected. Is there a reason that `clang-format-11` is still used?

omoerbeek commented:
> By the way, an unrelated question: I have tried to fix formatting in files I have touched, using the provided script (`./build-scripts/format-code...`). Doing that has produced wrong results, which failed on CI. On further inspection I saw that the script expects `clang-format-11`, while I had `clang-format-17` installed. After installing 11, it worked as expected. Is there a reason that `clang-format-11` is still used?

Version 11 was the one that was available in our CI when we started enforcing formatting. Later we noticed differences with newer clang-format versions, so we fixated the version. It mights be time to move to a newer version...

esensar commented:
I think everything is covered now. I still need to update documentation though.

rgacogne commented on pdns/dnsdistdist/dnsdist-lua-ffi.cc:
```diff
@@ -1086,7 +1085,7 @@ size_t dnsdist_ffi_dnsquestion_generate_proxy_protocol_payload(const dnsdist_ffi
     valuesVect.reserve(valuesCount);
     for (size_t idx = 0; idx < valuesCount; idx++) {
-      valuesVect.push_back({ std::string(values[idx].value, values[idx].size), values[idx].type });
+      valuesVect.push_back({std::string(values[idx].value, values[idx].size), values[idx].type});
```
```suggestion
      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
      valuesVect.push_back({std::string(values[idx].value, values[idx].size), values[idx].type});
```

rgacogne commented on pdns/dnsdistdist/dnsdist-lua.cc:
```diff
@@ -3410,44 +3413,92 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
   });
-  luaCtx.writeFunction("incMetric", [](const std::string& name, boost::optional<uint64_t> step) {
-    auto result = dnsdist::metrics::incrementCustomCounter(name, step ? *step : 1);
+  luaCtx.writeFunction("incMetric", [](const std::string& name, boost::optional<boost::variant<uint64_t, update_metric_opts_t>> opts) {
```
```suggestion
  // NOLINTNEXTLINE(performance-unnecessary-value-param)
  luaCtx.writeFunction("incMetric", [](const std::string& name, boost::optional<boost::variant<uint64_t, update_metric_opts_t>> opts) {
```

rgacogne commented on pdns/dnsdistdist/dnsdist-lua.cc:
```diff
@@ -3410,44 +3413,92 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
   });
-  luaCtx.writeFunction("decMetric", [](const std::string& name, boost::optional<uint64_t> step) {
-    auto result = dnsdist::metrics::decrementCustomCounter(name, step ? *step : 1);
+  luaCtx.writeFunction("decMetric", [](const std::string& name, boost::optional<boost::variant<uint64_t, update_metric_opts_t>> opts) {
```
```suggestion
  // NOLINTNEXTLINE(performance-unnecessary-value-param)
  luaCtx.writeFunction("decMetric", [](const std::string& name, boost::optional<boost::variant<uint64_t, update_metric_opts_t>> opts) {
```

rgacogne reviewed (commented):
Looks good to me, I suggested a few changes to appease the static analyzer. A new regression test would be nice, and of course as you already mentioned this needs to be documented :)

esensar commented:
I have added the documentation. I was not sure what to set as `versionAdded`. Also, should I have added variants with `options` as separate functions (in case of `incMetric`, `decMetric` and `declareMetric`) and deprecate old variants, or is this way of documenting arguments with different variants okay?

rgacogne commented on pdns/dnsdistdist/docs/reference/custommetrics.rst:
```diff
@@ -10,13 +10,19 @@ Then you can update those at runtime using the following functions, depending on
   .. versionchanged:: 1.8.1
     This function can now be used at runtime, instead of only at configuration time.
 
+  .. versionchanged:: ?
```
For now let's go with 2.0.0. I'll update the docs if this gets backported to 1.9.x.

rgacogne commented on pdns/dnsdistdist/docs/reference/custommetrics.rst:
```diff
@@ -25,51 +31,103 @@ Then you can update those at runtime using the following functions, depending on
+
+  Options:
+
+  * ``step``: int - By ow much the counter should be incremented, default to 1
```
```suggestion
  * ``step``: int - By how much the counter should be incremented, default to 1
```

rgacogne commented on pdns/dnsdistdist/docs/reference/custommetrics.rst:
```diff
@@ -25,51 +31,103 @@ Then you can update those at runtime using the following functions, depending on
-.. function:: getMetric(name) -> double
+  Options:
+
+  * ``step``: int - By ow much the counter should be decremented, default to 1
```
```suggestion
  * ``step``: int - By how much the counter should be decremented, default to 1
```

rgacogne reviewed (changes requested):
A couple nits but the documentation looks OK and the tests are quite nice, thanks!

rgacogne commented:
> I think for 2.0.0 (if breaking changes are allowed between major versions), we could just leave opts for `incMetric`, `decMetric` and `declareMetric`, instead of allowing previous optional parameter for these, but I guess that can be done in a separate PR for easier backporting.

Breaking changes in major versions are of course OK when they are needed or particularly useful, but my feeling is that it's neither in this case. The solution you came with is backward-compatible and quite clean :+1:.
